arangod/IResearch/ViewSnapshot.cpp

```diff
@@ -0,0 +1,180 @@
+///        the container storing the view state for a given TransactionState
+/// @note it is assumed that DBServer ViewState resides in the same
+///       TransactionState as the IResearchView ViewState, therefore a separate
+///       lock is not required to be held by the DBServer CompoundReader
+////////////////////////////////////////////////////////////////////////////////
+class ViewSnapshotCookie final : public ViewSnapshotImpl,
+                                 public TransactionState::Cookie {
+ public:
+  bool equalCollections(Links const& links);
+
+  void clear() noexcept;
+
+  void add(DataSourceId cid, irs::directory_reader&& reader) noexcept;
+
+  void set(Links&& links) noexcept;
+
+  std::vector<irs::directory_reader> _readers;
+  // prevent data-store deallocation (lock @ AsyncSelf)
+  Links _links;
+};
+
+bool ViewSnapshotCookie::equalCollections(Links const& links) {
+  if (_links.size() != links.size()) {
+    TRI_ASSERT(_links.empty());
+    return false;
+  }
+  for (auto const& entry : _links) {
+    if (!links.contains(entry.first)) {
+      // key not found -> maybe sync + create + return new key
+      // key     found -> return key
+      //             \--> sync + recompute + return key
+      TRI_ASSERT(false);
+      return false;
+    }
+  }
+  return true;
+}
+
+void ViewSnapshotCookie::clear() noexcept {
+  _live_docs_count = 0;
+  _docs_count = 0;
+  _segments.clear();
+  _readers.clear();
+  _links = {};
+}
+
+void ViewSnapshotCookie::add(DataSourceId cid,
+                             irs::directory_reader&& reader) noexcept {
+  _readers.reserve(_readers.size() + reader.size());
+  for (auto const& segment : reader) {
+    _segments.emplace_back(cid, &segment);
+  }
+  _live_docs_count += reader.live_docs_count();
+  _docs_count += reader.docs_count();
+  _readers.emplace_back(std::move(reader));
+}
+
+void ViewSnapshotCookie::set(Links&& links) noexcept {
+  _links = std::move(links);
+}
+
+}  // namespace
+
+ViewSnapshotView::ViewSnapshotView(
+    const ViewSnapshot& rhs,  // TODO(MBkkt) Maybe we should move?
+    containers::FlatHashSet<DataSourceId> const& collections) noexcept {
+  for (std::size_t i = 0, size = rhs.size(); i != size; ++i) {
+    auto const cid = rhs.cid(i);
+    if (!collections.contains(cid)) {
+      continue;
+    }
+    auto const& segment = rhs[i];
+    _docs_count += segment.docs_count();
+    _live_docs_count += segment.live_docs_count();
+    _segments.emplace_back(cid, &segment);
+  }
+}
+
+ViewSnapshot const* makeViewSnapshot(transaction::Methods& trx,
+                                     ViewSnapshotMode mode,
+                                     ViewSnapshot::Links&& links,
+                                     void const* key,
+                                     std::string_view name) noexcept {
+  if (links.empty()) {
+    static const ViewSnapshotImpl empty;
+    return &empty;  // TODO(MBkkt) Maybe nullptr?
```

Comment:
I'd say empty here is the right choice. View without links is not an error in general. It is just empty View/